invenio_records_permissions/generators.py

```diff
@@ -6,227 +6,192 @@
 # and/or modify it under the terms of the MIT License; see LICENSE file for
 # more details.
 
+"""Invenio Records Permissions Generators."""
+
 import json
 
 from elasticsearch_dsl.query import Q
 from flask import g
+from flask_login import current_user
 from flask_principal import ActionNeed, UserNeed
 from invenio_access.permissions import any_user
 from invenio_files_rest.models import Bucket, ObjectVersion
 from invenio_records_files.api import Record
 from invenio_records_files.models import RecordsBuckets
 
 
-class _NeedClass(object):
-
-    def needs(self):
-        pass
-
-    def excludes(self):
-        pass
+class Generator(object):
+    """Parent class mapping the context when an action is allowed or denied.
 
-    def query_filter(self):
-        pass
+    It does so by *generating* "needed" and "excluded" Needs. At the search
+    level it implements the *query filters* to restrict the search.
 
+    Any context inherits from this class.
+    """
 
-class _AnyUser(_NeedClass):
+    def needs(self, **kwargs):
+        """Enabling low-level Needs."""
+        return []
 
-    def __init__(self):
-        super(_AnyUser, self).__init__()
-
-    def needs(self):
-        return [any_user]
+    def excludes(self, **kwargs):
+        """Preventing low-level Needs."""
+        return []
 
-    def query_filter(self):
-        return Q('match_all')  # match all query
+    def query_filter(self, **kwargs):
+        """Elasticsearch filters."""
+        return []
 
 
-AnyUser = _AnyUser()
-
-
-class _Deny(_NeedClass):
+class AnyUser(Generator):
+    """Allows any user."""
 
     def __init__(self):
-        super(_Deny, self).__init__()
+        """Constructor."""
+        super(AnyUser, self).__init__()
 
-    def excludes(self):
+    def needs(self, **kwargs):
+        """Enabling low-level Needs."""
         return [any_user]
 
-    def query_filter(self):
-        return ~Q('match_all')  # Match None
-
+    def query_filter(self, **kwargs):
+        """Match all in search."""
+        return Q('match_all')
 
-Deny = _Deny()
 
-
-class _Admin(_NeedClass):
+class Deny(Generator):
+    """Denies ALL users (except super users)."""
```

Comment:
This is not fully correct, as per implementation. In addition Lars mentioned that Deny all should deny even superusers. Although now I am not sure if its the right way to go.